sparse/tpls/KokkosSparse_spmv_bsrmatrix_tpl_spec_decl.hpp

```diff
@@ -813,4 +814,231 @@ KOKKOSSPARSE_SPMV_MV_CUSPARSE(Kokkos::complex<float>, int, int,
+  using rocsparse_value_type =
+      typename KokkosSparse::Impl::kokkos_to_rocsparse_type<value_type>::type;
+
+  // assert ordinals and offsets are the expected types
+  static_assert(std::is_same_v<offset_type, rocsparse_int>,
+                "A offset_type must be rocsparse_int");
+  static_assert(std::is_same_v<ordinal_type, rocsparse_int>,
+                "A ordinal_type must be rocsparse_int");
+
+  // assert all operands are the same type
+  using x_value_type = typename XVector::non_const_value_type;
+  using y_value_type = typename YVector::non_const_value_type;
+  static_assert(std::is_same_v<value_type, x_value_type>,
+                "A and x must have same value type");
+  static_assert(std::is_same_v<value_type, y_value_type>,
+                "A and y must have same value type");
+
+  // assert X and Y are non-stride (pass raw pointers to TPL)
+  static_assert(
+      !std::is_same_v<typename XVector::array_layout, Kokkos::LayoutStride>,
+      "x must be contiguous");
+  static_assert(
+      !std::is_same_v<typename YVector::array_layout, Kokkos::LayoutStride>,
+      "y must be contiguous");
+
+  // assert BSR data is non-stride (pass raw pointers to TPL)
+  static_assert(!std::is_same_v<typename AMatrix::values_type::array_layout,
+                                Kokkos::LayoutStride>,
+                "A values must be contiguous");
+  static_assert(!std::is_same_v<typename AMatrix::row_map_type::array_layout,
+                                Kokkos::LayoutStride>,
+                "A row_map must be contiguous");
+  static_assert(!std::is_same_v<typename AMatrix::index_type::array_layout,
+                                Kokkos::LayoutStride>,
+                "A entries must be contiguous");
+
+  rocsparse_handle handle = controls.getRocsparseHandle();
+
+  // set the mode
+  rocsparse_operation trans;
+  switch (toupper(mode[0])) {
+    case 'N': trans = rocsparse_operation_none; break;
+    default: {
+      std::stringstream ss;
+      ss << "Mode " << mode << " invalid for rocsparse_[*]bsrmv\n";
+      throw std::invalid_argument(ss.str());
+    }
+  }
+
+  /*
+  Specify the matrix direction.
+  The rocsparse_direction indicates whether a dense matrix should be parsed by
+  rows or by columns, assuming column-major storage. Values: enumerator
+  rocsparse_direction_row Parse the matrix by rows. enumerator
+  rocsparse_direction_column Parse the matrix by columns.
+  */
+  // KokkosSparse Bsr matrix blocks are layoutright (row-major)
+  static_assert(
+      std::is_same_v<typename AMatrix::block_layout, Kokkos::LayoutRight>,
+      "A blocks must be stored layout-right");
+  rocsparse_direction dir = rocsparse_direction_row;
+
+  const rocsparse_int mb = rocsparse_int(A.numRows());  // number of block rows
+  const rocsparse_int nb = rocsparse_int(A.numCols());  // number of block cols
+  const rocsparse_int nnzb = rocsparse_int(
+      A.nnz() / A.blockDim() / A.blockDim());  // number of non-zero blocks
```

Comment:
A is a BsrMatrix, right? I think just A.nnz() is the number of nonzero blocks?
```
 885   //! The number of stored entries in the sparse matrix.
 886   KOKKOS_INLINE_FUNCTION size_type nnz() const {
 887     return graph.entries.extent(0);
 888   }
```

Reply:
You're right, and this is not caught by the existing Bsr unit tests, so I will overhaul them.